File: Project_mbs/userfctR/user_JointForces.py
# -*- coding: utf-8 -*-
"""Module for the definition of joint forces."""
# Author: Robotran Team
# (c) Universite catholique de Louvain, 2020
import numpy as np

def user_JointForces(mbs_data, tsim):
    """Compute the force and torques in the joint.

    It fills the MBsysPy.MbsData.Qq array.

    Parameters
    ----------
    mbs_data : MBsysPy.MbsData
        The multibody system associated to this computation.
    tsim : float
        The current time of the simulation.

    Notes
    -----
    The numpy.ndarray MBsysPy.MbsData.Qq is 1D array with index starting at 1.
    The first index (array[0]) must not be modified. The first index to be
    filled is array[1].

    Returns
    -------
    None
    """
    
    mbs_data.Qq[1:] = 0.
    motor_ids = [mbs_data.joint_id["R_lh"],mbs_data.joint_id["R_rh"] ,mbs_data.joint_id["R_lk"],mbs_data.joint_id["R_rk"]]
    q_motors  = [mbs_data.q[motor_ids[0]] , mbs_data.q[motor_ids[1]] ,mbs_data.q[motor_ids[2]] ,mbs_data.q[motor_ids[3]]]
    qd_motors = [mbs_data.qd[motor_ids[0]], mbs_data.qd[motor_ids[1]],mbs_data.qd[motor_ids[2]],mbs_data.qd[motor_ids[3]]]


    trajectory_index = int(tsim // mbs_data.frequency)
    q_ref = mbs_data.reference_trajectory[trajectory_index][1:]
    
    ### Dynamixel Controller ###
    Kp = 900.0/128
    PWM_goal = 885.0
    Nominal_voltage = 12.0
    
    PWM = (q_ref - q_motors) * (4095.0 / (2 * np.pi) * Kp)
    PWM_sat = np.clip(PWM, PWM_goal, PWM_goal)
    
    u = PWM_sat * (Nominal_voltage / PWM_goal)   
    
    ### Motor Equations ###

    return

File: philippides/userfctR/user_JointForces.c
/* ---------------------------
 * Robotran - MBsysC
 * 
 * Template file for the user JointForces function
 * 
 * (c) Universite catholique de Louvain
 *     
 */

#include "math.h" 

#include "mbs_data.h"
#include "user_model.h"
#include "set_output.h"
#include "useful_functions.h"

typedef struct {
    double **trajectory;  // 2D array storing the trajectory data
    int index;           // Index for tracking progress in the trajectory
} Trajectory;

// Function to initialize trajectory data by reading from a CSV file
Trajectory* init_trajectory(const char *filename, int rows, int cols) {
    Trajectory *traj = (Trajectory*)malloc(sizeof(Trajectory)); // Allocate memory for trajectory structure
    traj->trajectory = (double**)malloc(rows * sizeof(double*)); // Allocate memory for rows
    for (int i = 0; i < rows; i++) {
        traj->trajectory[i] = (double*)malloc(cols * sizeof(double)); // Allocate memory for columns
    }
    traj->index = 0;
    
    FILE *file = fopen(filename, "r"); // Open the file for reading
    if (!file) {
        perror("Error opening file");
        return NULL;
    }
    
    char buffer[1024];
    fgets(buffer, sizeof(buffer), file); // Skip the header line
    
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            fscanf(file, "%lf,", &traj->trajectory[i][j]); // Read values from file
        }
    }
    fclose(file); // Close the file
    return traj;
}

// Function to compute joint forces based on motor states and reference trajectory
double* user_JointForces(MbsData *mbs_data, double tsim) {
    // Reset all forces in the joint force array
    for (int i = 1; i < MAX_JOINTS; i++) {
        mbs_data->Qq[i] = 0.0;
    }
    
    // Retrieve motor IDs for different joints
    int motor_ids[4] = {mbs_data->joint_id["R_lh"], mbs_data->joint_id["R_rh"],
                         mbs_data->joint_id["R_lk"], mbs_data->joint_id["R_rk"]};
    
    // Retrieve motor positions and velocities
    double q_motors[4], qd_motors[4];
    for (int i = 0; i < 4; i++) {
        q_motors[i] = mbs_data->q[motor_ids[i]];
        qd_motors[i] = mbs_data->qd[motor_ids[i]];
    }
    
    // Compute trajectory index based on simulation time
    int trajectory_index = (int)(tsim * mbs_data->frequency);
    if (trajectory_index >= TOTAL_TRAJECTORY_ROWS) {
        trajectory_index = TOTAL_TRAJECTORY_ROWS - 1; // Prevent out-of-bounds access
    }
    
    // Get reference joint positions for the current time step
    double q_ref[4];
    for (int i = 0; i < 4; i++) {
        q_ref[i] = reference_trajectory->trajectory[trajectory_index][i+1];
    }
    
    // Controller parameters
    double Kp = 900.0 / 128.0;
    double PWM_goal = 885.0;
    double Nominal_voltage = 12.0;
    
    // Compute PWM signals for motor control
    double PWM[4], PWM_sat[4], u[4];
    for (int i = 0; i < 4; i++) {
        PWM[i] = (q_ref[i] - q_motors[i]) * (4095.0 / (2 * M_PI) * Kp);
        PWM_sat[i] = fmax(-PWM_goal, fmin(PWM[i], PWM_goal)); // Clipping function
        u[i] = PWM_sat[i] * (Nominal_voltage / PWM_goal);
    }
    
    // Store computed voltages for motors
    for (int i = 0; i < 4; i++) {
        mbs_data->Voltages[i] = u[i];
    }
    
    // Motor parameters
    double HGR = 353.5;   // Hip gear ratio
    double KGR = 212.6;   // Knee gear ratio
    double ktp = 0.395 / HGR;   // Torque constant w.r.t. voltage [Nm/V]
    double Kvp = 1.589 / (HGR * HGR); // Viscous friction constant [Nm*s/rad]
    double τc_u = 0.065 / HGR;  // Dry friction torque [Nm]
    
    // Compute motor torques
    double gear_ratios[4] = {HGR, HGR, KGR, KGR};
    double w[4], τ_0[4], τ_m[4];
    for (int i = 0; i < 4; i++) {
        w[i] = qd_motors[i] * gear_ratios[i];
        τ_0[i] = u[i] * gear_ratios[i] * ktp - w[i] * gear_ratios[i] * Kvp;
        τ_m[i] = τ_0[i] - ((w[i] > 0) ? τc_u * gear_ratios[i] : -τc_u * gear_ratios[i]);
    }
    
    // Assign computed torques to joint force array
    for (int i = 0; i < 4; i++) {
        mbs_data->Qq[motor_ids[i]] = τ_m[i];
    }
    
    return mbs_data->Qq;
}
